enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/tx/edge/TxPollingClient.java

```diff
@@ -19,72 +19,181 @@
  */
 package org.neo4j.coreedge.catchup.tx.edge;
 
-import java.util.function.Supplier;
-
 import org.neo4j.coreedge.catchup.storecopy.CoreClient;
+import org.neo4j.coreedge.raft.RenewableTimeoutService;
+import org.neo4j.coreedge.raft.RenewableTimeoutService.RenewableTimeout;
+import org.neo4j.coreedge.raft.RenewableTimeoutService.TimeoutName;
 import org.neo4j.coreedge.server.AdvertisedSocketAddress;
 import org.neo4j.coreedge.server.edge.CoreServerSelectionStrategy;
-import org.neo4j.kernel.impl.transaction.log.TransactionIdStore;
-import org.neo4j.kernel.impl.util.JobScheduler;
 import org.neo4j.kernel.lifecycle.LifecycleAdapter;
 import org.neo4j.logging.Log;
 import org.neo4j.logging.LogProvider;
 
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
-
-import static org.neo4j.kernel.impl.util.JobScheduler.Groups.pullUpdates;
+import static org.neo4j.coreedge.catchup.tx.edge.TxPollingClient.States.INACTIVE;
+import static org.neo4j.coreedge.catchup.tx.edge.TxPollingClient.States.RECEIVING_TRANSACTIONS;
+import static org.neo4j.coreedge.catchup.tx.edge.TxPollingClient.States.WAIT_NEXT_PULL;
+import static org.neo4j.coreedge.catchup.tx.edge.TxPollingClient.Timeouts.TX_PULLER_TIMEOUT;
 
-public class TxPollingClient extends LifecycleAdapter
+/**
+ * This class is responsible for pulling transactions from a core server and queuing
+ * them to be applied with the {@link BatchingTxApplier}. It operates in two main states:
+ *
+ *   WAIT_NEXT_PULL:        Waiting for the next "tick" for pulling transactions.
+ *   RECEIVING_TRANSACTION: Receiving transactions during the pull.
+ *
+ * There is also the initial INACTIVE state, also entered when stopping.
+ *
+ * Pull requests are issued on a fixed interval, but skipped if the {@link BatchingTxApplier}
+ * isn't yet finished with the current work.
+ */
+public class TxPollingClient extends LifecycleAdapter implements TxPullListener
 {
-    private final JobScheduler jobScheduler;
-    private final Supplier<TransactionIdStore> transactionIdStoreSupplier;
-    private final CoreServerSelectionStrategy connectionStrategy;
-    private final long pollingInterval;
-    private final CoreClient coreClient;
-    private final TxPullResponseListener txPullResponseListener;
+    enum Timeouts implements TimeoutName
+    {
+        TX_PULLER_TIMEOUT
+    }
+
+    interface Handler
+    {
+        default void onTimeout( TxPollingClient ctx ) {}
+        default void onTxReceived( TxPollingClient ctx, TxPullResponse tx ) {}
+        default void onTxStreamingComplete( TxPollingClient ctx ) {}
```

Comment:
No, this is an interface which also defines the default behaviour of not doing anything for simplicity.